dvc/parsing/__init__.py

```diff
@@ -499,8 +499,26 @@ def _resolve_matrix_data(self) -> Dict[str, List]:
             iterable = self.context.resolve(self.matrix_data, unwrap=False)
         except (ContextError, ParseError) as exc:
             format_and_raise(exc, f"'{self.where}.{self.name}.matrix'", self.relpath)
+
+        # Matrix entries will have `key` and `item` added to the context.
+        # Warn users if these are already in the context from the global vars.
+        self._warn_if_overwriting(["key", "item"])
```

Comment:
Let's not hardcode these here.

```suggestion
        self._warn_if_overwriting([self.pair, self.value])
```